### PR TITLE
Implement ls

### PR DESCRIPTION
**Pull request by roblabla**
EzPz.

### CODE REVIEW

File: shell/src/main.rs [hunk 1]
```diff
@@ -254,6 +258,47 @@ fn cat<W: Write>(f: &mut W, filesystem: &IFileSystemProxy, file: &str) -> Result
...
+    let path = path.unwrap_or(".");
+    let path = get_path_relative_to_current_directory(path);
+    if path.len() > 0x300 {
+        return Err(FileSystemError::InvalidInput.into())
+    }
+
+    let mut ipc_path = [0x0; 0x300];
+    ipc_path[..path.as_bytes().len()].copy_from_slice(path.as_bytes());
+
+    let directory = filesystem.open_directory(3, &ipc_path)?;
```
Comment (Orycterope) [inline shell/src/main.rs]: what about 
```
ls some_file
```
 ?

File: shell/src/main.rs [hunk 2]
```diff
@@ -254,6 +258,47 @@ fn cat<W: Write>(f: &mut W, filesystem: &IFileSystemProxy, file: &str) -> Result
     Ok(())
 }
 
+/// List files and folders at the given path, or in the current path is none is
+/// given.
+fn ls(mut terminal: &mut Terminal, filesystem: &IFileSystemProxy, path: Option<&str>) -> Result<(), Error> {
+    use sunrise_libuser::fs::{DirectoryEntry, DirectoryEntryType};
+
+    let path = path.unwrap_or(".");
+    let path = get_path_relative_to_current_directory(path);
```
Comment (Orycterope) [inline shell/src/main.rs]: does this work for absolute paths ?

File: shell/src/main.rs [hunk 3]
```diff
@@ -254,6 +258,47 @@ fn cat<W: Write>(f: &mut W, filesystem: &IFileSystemProxy, file: &str) -> Result
...
+        for entry in entries {
+            let split_at = entry.path.iter().position(|v| *v == 0).unwrap_or(0x300);
+            let prefix_len = if path == "/" {
+                // The prefix to remove is literally just the leading `/`
+                1
+            } else {
+                // The prefix to remove is the path + the trailing `/`.
+                path.len() + 1
+            };
+            let _ = writeln!(&mut terminal, "{}", String::from_utf8_lossy(&entry.path[prefix_len..split_at]));
```
[the remark this answers is not in the document]
Comment (Orycterope) [reply shell/src/main.rs]: Or color in green, the logger can do it IIRC

File: shell/src/main.rs (hunk 2 above)
Comment (roblabla) [reply shell/src/main.rs]: I checked, and the answer is yes.

File: shell/src/main.rs (hunk 1 above)
Comment (roblabla) [reply shell/src/main.rs]: Currently returns an error saying it's not a directory. Will fix tomorrow.

File: shell/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (roblabla) [reply shell/src/main.rs]: Colors is probably a better idea. Will take a look tomorrow, I think Terminal can output some colors.

**Pull request merged**